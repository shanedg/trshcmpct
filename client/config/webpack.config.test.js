const ESLintWebpackPlugin = require('eslint-webpack-plugin');
const HtmlWebpackPlugin = require('html-webpack-plugin');
const serializer = require('jest-serializer-path');
const webpack = require('webpack');

const webpackConfig = require('./webpack.config');

// Remove absolute file paths from snapshots.
expect.addSnapshotSerializer(serializer);

describe('webpack', () => {
  const developmentEnvironment = {};
  const productionEnvironment = { production: true };
  let developmentConfig;
  let productionConfig;

  describe('common config', () => {
    beforeAll(() => {
      developmentConfig = webpackConfig(developmentEnvironment);
      productionConfig = webpackConfig(productionEnvironment);
      Object.freeze(developmentConfig);
      Object.freeze(productionConfig);
    });

    it('sets plugins', () => {
      expect(developmentConfig.plugins).toMatchObject([
        expect.any(ESLintWebpackPlugin),
        expect.any(webpack.DefinePlugin),
        expect.any(HtmlWebpackPlugin),
      ]);
      expect(productionConfig.plugins).toMatchObject([
        expect.any(ESLintWebpackPlugin),
        expect.any(webpack.DefinePlugin),
        expect.any(HtmlWebpackPlugin),
      ]);
    });
  });

  describe('development config', () => {
    beforeAll(() => {
      developmentConfig = webpackConfig(developmentEnvironment);
      Object.freeze(developmentConfig);
    });

    it('matches snapshot', () => {
      expect(developmentConfig).toMatchSnapshot();
    });

    it('sets development mode', () => {
      expect(developmentConfig.mode).toEqual('development');
    });

    it('turns off minimize optimization', () => {
      expect(developmentConfig.optimization).toHaveProperty('minimize', false);
    });

    it('DefinePlugin sets __DEV__ to "true"', () => {
      expect(developmentConfig.plugins.find(p => p instanceof webpack.DefinePlugin))
        .toHaveProperty('definitions.__DEV__', 'true');
    });
  });

  describe('production config', () => {
    beforeAll(() => {
      productionConfig = webpackConfig(productionEnvironment);
      Object.freeze(productionConfig);
    });

    it('matches snapshot', () => {
      expect(productionConfig).toMatchSnapshot();
    });

    it('sets production mode', () => {
      expect(productionConfig.mode).toEqual('production');
    });

    it('turns on minimize optimization', () => {
      expect(productionConfig.optimization).toHaveProperty('minimize', true);
    });

    it('DefinePlugin set __DEV__ to "false"', () => {
      expect(productionConfig.plugins.find(p => p instanceof webpack.DefinePlugin))
        .toHaveProperty('definitions.__DEV__', 'false');
    });
  });
});
