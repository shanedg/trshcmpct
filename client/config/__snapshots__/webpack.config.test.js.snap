// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`webpack 0 - webapp development config matches snapshot 1`] = `
Object {
  "devtool": "eval-source-map",
  "entry": Object {
    "bootstrap": "<PROJECT_ROOT>/src/bootstrap.ts",
  },
  "mode": "development",
  "module": Object {
    "rules": Array [
      Object {
        "exclude": /node_modules/,
        "loader": "babel-loader",
        "options": Object {
          "configFile": "<PROJECT_ROOT>/config/babel.config.cjs",
        },
        "test": /\\\\\\.\\(js\\|jsx\\|ts\\|tsx\\)\\$/,
      },
      Object {
        "test": /\\\\\\.css\\$/i,
        "use": Array [
          "style-loader",
          "css-loader",
        ],
      },
    ],
  },
  "name": "webapp",
  "optimization": Object {
    "minimize": false,
    "minimizer": Array [
      TerserPlugin {
        "options": Object {
          "exclude": undefined,
          "extractComments": true,
          "include": undefined,
          "minimizer": Object {
            "implementation": [Function],
            "options": Object {
              "warnings": false,
            },
          },
          "parallel": true,
          "test": /\\\\\\.\\[cm\\]\\?js\\(\\\\\\?\\.\\*\\)\\?\\$/i,
        },
      },
    ],
    "splitChunks": Object {
      "cacheGroups": Object {
        "defaultVendors": Object {
          "priority": -10,
          "test": /\\[\\\\\\\\/\\]node_modules\\[\\\\\\\\/\\]/,
        },
      },
      "chunks": "async",
      "minChunks": 1,
      "minSize": 30000,
    },
  },
  "output": Object {
    "clean": true,
    "filename": "[name].[chunkhash].js",
    "path": "<PROJECT_ROOT>/dist",
  },
  "plugins": Array [
    ESLintWebpackPlugin {
      "key": "ESLintWebpackPlugin",
      "options": Object {
        "cache": true,
        "cacheLocation": "node_modules/.cache/eslint-cache/",
        "emitError": false,
        "emitWarning": true,
        "extensions": Array [
          ".js",
          ".jsx",
          ".ts",
          ".tsx",
        ],
        "failOnError": false,
        "lintDirtyModulesOnly": false,
        "reportUnusedDisableDirectives": "warn",
        "resourceQueryExclude": Array [],
      },
      "run": [Function],
    },
    WebpackManifestPlugin {
      "options": Object {
        "assetHookStage": Infinity,
        "basePath": "",
        "fileName": "manifest.json",
        "filter": null,
        "generate": undefined,
        "map": null,
        "publicPath": "",
        "removeKeyHash": /\\(\\[a-f0-9\\]\\{16,32\\}\\\\\\.\\?\\)/gi,
        "seed": undefined,
        "serialize": [Function],
        "sort": null,
        "transformExtensions": /\\^\\(gz\\|map\\)\\$/i,
        "useEntryKeys": false,
        "useLegacyEmit": false,
        "writeToFileEmit": false,
      },
    },
    HtmlWebpackPlugin {
      "userOptions": Object {
        "template": "<PROJECT_ROOT>/src/index.html",
        "title": "trshcmpctr",
      },
      "version": 5,
    },
  ],
  "resolve": Object {
    "extensions": Array [
      ".js",
      ".jsx",
      ".ts",
      ".tsx",
    ],
  },
}
`;

exports[`webpack 0 - webapp production config matches snapshot 1`] = `
Object {
  "devtool": "source-map",
  "entry": Object {
    "bootstrap": "<PROJECT_ROOT>/src/bootstrap.ts",
  },
  "mode": "production",
  "module": Object {
    "rules": Array [
      Object {
        "exclude": /node_modules/,
        "loader": "babel-loader",
        "options": Object {
          "configFile": "<PROJECT_ROOT>/config/babel.config.cjs",
        },
        "test": /\\\\\\.\\(js\\|jsx\\|ts\\|tsx\\)\\$/,
      },
      Object {
        "test": /\\\\\\.css\\$/i,
        "use": Array [
          "style-loader",
          "css-loader",
        ],
      },
    ],
  },
  "name": "webapp",
  "optimization": Object {
    "minimize": true,
    "minimizer": Array [
      TerserPlugin {
        "options": Object {
          "exclude": undefined,
          "extractComments": true,
          "include": undefined,
          "minimizer": Object {
            "implementation": [Function],
            "options": Object {
              "warnings": false,
            },
          },
          "parallel": true,
          "test": /\\\\\\.\\[cm\\]\\?js\\(\\\\\\?\\.\\*\\)\\?\\$/i,
        },
      },
    ],
    "splitChunks": Object {
      "cacheGroups": Object {
        "defaultVendors": Object {
          "priority": -10,
          "test": /\\[\\\\\\\\/\\]node_modules\\[\\\\\\\\/\\]/,
        },
      },
      "chunks": "async",
      "minChunks": 1,
      "minSize": 30000,
    },
  },
  "output": Object {
    "clean": true,
    "filename": "[name].[chunkhash].js",
    "path": "<PROJECT_ROOT>/dist",
  },
  "plugins": Array [
    ESLintWebpackPlugin {
      "key": "ESLintWebpackPlugin",
      "options": Object {
        "cache": true,
        "cacheLocation": "node_modules/.cache/eslint-cache/",
        "emitError": true,
        "emitWarning": false,
        "extensions": Array [
          ".js",
          ".jsx",
          ".ts",
          ".tsx",
        ],
        "failOnError": true,
        "lintDirtyModulesOnly": false,
        "reportUnusedDisableDirectives": null,
        "resourceQueryExclude": Array [],
      },
      "run": [Function],
    },
    WebpackManifestPlugin {
      "options": Object {
        "assetHookStage": Infinity,
        "basePath": "",
        "fileName": "manifest.json",
        "filter": null,
        "generate": undefined,
        "map": null,
        "publicPath": "",
        "removeKeyHash": /\\(\\[a-f0-9\\]\\{16,32\\}\\\\\\.\\?\\)/gi,
        "seed": undefined,
        "serialize": [Function],
        "sort": null,
        "transformExtensions": /\\^\\(gz\\|map\\)\\$/i,
        "useEntryKeys": false,
        "useLegacyEmit": false,
        "writeToFileEmit": false,
      },
    },
    HtmlWebpackPlugin {
      "userOptions": Object {
        "template": "<PROJECT_ROOT>/src/index.html",
        "title": "trshcmpctr",
      },
      "version": 5,
    },
  ],
  "resolve": Object {
    "extensions": Array [
      ".js",
      ".jsx",
      ".ts",
      ".tsx",
    ],
  },
}
`;

exports[`webpack 1 - library-module development config matches snapshot 1`] = `
Object {
  "entry": Object {
    "index": "<PROJECT_ROOT>/src/index.ts",
  },
  "experiments": Object {
    "outputModule": true,
  },
  "mode": "development",
  "module": Object {
    "rules": Array [
      Object {
        "exclude": /node_modules/,
        "loader": "babel-loader",
        "options": Object {
          "configFile": "<PROJECT_ROOT>/config/babel.config.cjs",
        },
        "test": /\\\\\\.\\(js\\|jsx\\|ts\\|tsx\\)\\$/,
      },
      Object {
        "test": /\\\\\\.css\\$/i,
        "use": Array [
          "style-loader",
          "css-loader",
        ],
      },
    ],
  },
  "name": "library-module",
  "optimization": Object {
    "minimize": false,
    "minimizer": Array [
      TerserPlugin {
        "options": Object {
          "exclude": undefined,
          "extractComments": true,
          "include": undefined,
          "minimizer": Object {
            "implementation": [Function],
            "options": Object {
              "warnings": false,
            },
          },
          "parallel": true,
          "test": /\\\\\\.\\[cm\\]\\?js\\(\\\\\\?\\.\\*\\)\\?\\$/i,
        },
      },
    ],
    "splitChunks": Object {
      "cacheGroups": Object {
        "defaultVendors": Object {
          "priority": -10,
          "test": /\\[\\\\\\\\/\\]node_modules\\[\\\\\\\\/\\]/,
        },
      },
      "chunks": "async",
      "minChunks": 1,
      "minSize": 30000,
    },
  },
  "output": Object {
    "clean": true,
    "filename": "[name].js",
    "library": Object {
      "type": "module",
    },
    "module": true,
    "path": "<PROJECT_ROOT>/lib",
  },
  "plugins": Array [
    ESLintWebpackPlugin {
      "key": "ESLintWebpackPlugin",
      "options": Object {
        "cache": true,
        "cacheLocation": "node_modules/.cache/eslint-cache/",
        "emitError": false,
        "emitWarning": true,
        "extensions": Array [
          ".js",
          ".jsx",
          ".ts",
          ".tsx",
        ],
        "failOnError": false,
        "lintDirtyModulesOnly": false,
        "reportUnusedDisableDirectives": "warn",
        "resourceQueryExclude": Array [],
      },
      "run": [Function],
    },
  ],
  "resolve": Object {
    "extensions": Array [
      ".js",
      ".jsx",
      ".ts",
      ".tsx",
    ],
  },
  "target": "node16",
}
`;

exports[`webpack 1 - library-module production config matches snapshot 1`] = `
Object {
  "entry": Object {
    "index": "<PROJECT_ROOT>/src/index.ts",
  },
  "experiments": Object {
    "outputModule": true,
  },
  "mode": "production",
  "module": Object {
    "rules": Array [
      Object {
        "exclude": /node_modules/,
        "loader": "babel-loader",
        "options": Object {
          "configFile": "<PROJECT_ROOT>/config/babel.config.cjs",
        },
        "test": /\\\\\\.\\(js\\|jsx\\|ts\\|tsx\\)\\$/,
      },
      Object {
        "test": /\\\\\\.css\\$/i,
        "use": Array [
          "style-loader",
          "css-loader",
        ],
      },
    ],
  },
  "name": "library-module",
  "optimization": Object {
    "minimize": true,
    "minimizer": Array [
      TerserPlugin {
        "options": Object {
          "exclude": undefined,
          "extractComments": true,
          "include": undefined,
          "minimizer": Object {
            "implementation": [Function],
            "options": Object {
              "warnings": false,
            },
          },
          "parallel": true,
          "test": /\\\\\\.\\[cm\\]\\?js\\(\\\\\\?\\.\\*\\)\\?\\$/i,
        },
      },
    ],
    "splitChunks": Object {
      "cacheGroups": Object {
        "defaultVendors": Object {
          "priority": -10,
          "test": /\\[\\\\\\\\/\\]node_modules\\[\\\\\\\\/\\]/,
        },
      },
      "chunks": "async",
      "minChunks": 1,
      "minSize": 30000,
    },
  },
  "output": Object {
    "clean": true,
    "filename": "[name].js",
    "library": Object {
      "type": "module",
    },
    "module": true,
    "path": "<PROJECT_ROOT>/lib",
  },
  "plugins": Array [
    ESLintWebpackPlugin {
      "key": "ESLintWebpackPlugin",
      "options": Object {
        "cache": true,
        "cacheLocation": "node_modules/.cache/eslint-cache/",
        "emitError": true,
        "emitWarning": false,
        "extensions": Array [
          ".js",
          ".jsx",
          ".ts",
          ".tsx",
        ],
        "failOnError": true,
        "lintDirtyModulesOnly": false,
        "reportUnusedDisableDirectives": null,
        "resourceQueryExclude": Array [],
      },
      "run": [Function],
    },
  ],
  "resolve": Object {
    "extensions": Array [
      ".js",
      ".jsx",
      ".ts",
      ".tsx",
    ],
  },
  "target": "node16",
}
`;
